File: app/src/main/java/org/nevack/unitconverter/converter/ConverterDisplayView.java
package org.nevack.unitconverter.converter;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Rect;
import android.os.Build;

import androidx.annotation.Nullable;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import androidx.core.content.ContextCompat;
import androidx.core.text.HtmlCompat;

import android.text.Html;
import android.text.Spanned;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroupOverlay;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.SpinnerAdapter;
import android.widget.TextView;

import org.nevack.unitconverter.R;
import org.nevack.unitconverter.converter.ConverterContract.ConvertData;
import org.nevack.unitconverter.model.Unit;

import java.util.List;

import static androidx.core.text.HtmlCompat.FROM_HTML_MODE_LEGACY;

public class ConverterDisplayView extends LinearLayout {

    private final Spinner sourceSpinner;
    private final EditText sourceEditText;
    private final TextView sourceTextView;
    private final EditText resultEditText;
    private final TextView resultTextView;
    private final Spinner resultSpinner;
    private final FloatingActionButton fab;

    private static List<Unit> units;

    public ConverterDisplayView(final Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        setOrientation(VERTICAL);

        inflate(context, R.layout.display, this);

        fab = findViewById(R.id.fab);

        sourceTextView = findViewById(R.id.source_value_symbol);
        resultTextView = findViewById(R.id.result_value_symbol);

        sourceSpinner = findViewById(R.id.source_spinner);
        resultSpinner = findViewById(R.id.result_spinner);

        sourceEditText = findViewById(R.id.source_value);
        resultEditText = findViewById(R.id.result_value);

        fab.setOnClickListener(v -> {
            String temp = sourceEditText.getText().toString();
            sourceEditText.setText(resultEditText.getText().toString());
            resultEditText.setText(temp);
            int position = sourceSpinner.getSelectedItemPosition();
            sourceSpinner.setSelection(resultSpinner.getSelectedItemPosition());
            resultSpinner.setSelection(position);
        });

        sourceTextView.setOnClickListener(v -> sourceSpinner.performClick());

        resultTextView.setOnClickListener(v -> resultSpinner.performClick());
    }

    public void setTextWatcher(TextWatcher watcher) {
        sourceEditText.addTextChangedListener(watcher);
    }

    public void showResult(String result) {
        resultEditText.setText(result);
    }

    public ConvertData getConvertData() {
        return new ConvertData(
                sourceEditText.getText().toString(),
                resultEditText.getText().toString(),
                sourceSpinner.getSelectedItemPosition(),
                resultSpinner.getSelectedItemPosition()
        );
    }

    public String getCopyResult(boolean withUnitSymbols) {
        return resultEditText.getText().toString()
                + (withUnitSymbols ? resultTextView.getText().toString() : "");
    }

    public void showError() {
        resultEditText.setText(R.string.message_error);
    }

    private void setSpinnerAdapter(SpinnerAdapter adapter) {
        sourceSpinner.setAdapter(adapter);
        resultSpinner.setAdapter(adapter);
    }

    public void setUnits(List<Unit> units) {
        ConverterDisplayView.units = units;
        ArrayAdapter<String> adapter =
                new ArrayAdapter<>(getContext(), android.R.layout.simple_list_item_1);
        for (Unit unit : ConverterDisplayView.units) {
            adapter.add(unit.getName());
        }

        setSpinnerAdapter(adapter);
    }

    public void erase() {
        ViewGroupOverlay groupOverlay;
        View revealView = new View(getContext());
        Animator revealAnimator;

        final Rect displayRect = new Rect();
        this.getGlobalVisibleRect(displayRect);

        // Make reveal cover the display and status bar.
        revealView.setBottom(displayRect.bottom);
        revealView.setLeft(displayRect.left);
        revealView.setRight(displayRect.right);
        revealView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.colorAccent));

        groupOverlay = (ViewGroupOverlay) ((Activity) getContext()).getWindow().getDecorView().getOverlay();
        groupOverlay.add(revealView);

        final int[] clearLocation = new int[2];
        this.getLocationInWindow(clearLocation);
        clearLocation[0] += this.getWidth();
        clearLocation[1] += this.getHeight();

        final int revealCenterX = clearLocation[0] - revealView.getLeft();
        final int revealCenterY = clearLocation[1] - revealView.getTop();

        final double x1_2 = Math.pow(revealView.getLeft() - revealCenterX, 2);
        final double x2_2 = Math.pow(revealView.getRight() - revealCenterX, 2);
        final double y_2 = Math.pow(revealView.getTop() - revealCenterY, 2);
        final float revealRadius = (float) Math.max(Math.sqrt(x1_2 + y_2), Math.sqrt(x2_2 + y_2));

        revealAnimator = ViewAnimationUtils.createCircularReveal(revealView,
                revealCenterX, revealCenterY, 0.0f, revealRadius);

        revealAnimator.setDuration(
                getResources().getInteger(android.R.integer.config_longAnimTime));

        final Animator alphaAnimator = ObjectAnimator.ofFloat(revealView, View.ALPHA, 0.0f);
        alphaAnimator.setDuration(
                getResources().getInteger(android.R.integer.config_mediumAnimTime));
        alphaAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                clear();
            }
        });

        final AnimatorSet animatorSet = new AnimatorSet();

        //Play Reveal Animation if Lollipop or higher, and only alpha animation for others
        animatorSet.play(revealAnimator).before(alphaAnimator);

        animatorSet.setInterpolator(new AccelerateDecelerateInterpolator());

        final ViewGroupOverlay finalGroupOverlay = groupOverlay;
        final View finalRevealView = revealView;

        animatorSet.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animator) {
                finalGroupOverlay.remove(finalRevealView);
            }
        });

        animatorSet.start();
    }

    public void clear() {
        resultEditText.getEditableText().clear();
        sourceEditText.getEditableText().clear();
    }

    public void appendText(String text) {
        sourceEditText.append(text);
    }

    public void removeLastDigit() {
        String textString = sourceEditText.getText().toString();
        if (textString.length() > 0) {
            sourceEditText.setText(textString.substring(0, textString.length() - 1));
            sourceEditText.setSelection(sourceEditText.getText().length());
        }
    }

    public void setSpinnersCallback(SpinnerListener.Callback callback) {
        resultSpinner.setOnItemSelectedListener(new SpinnerListener(resultTextView, callback));
        sourceSpinner.setOnItemSelectedListener(new SpinnerListener(sourceTextView, callback));
    }

    public void setConvertData(ConvertData convertData) {
        sourceEditText.setText(convertData.getValue());
        resultEditText.setText(convertData.getResult());
        sourceSpinner.setSelection(convertData.getFrom());
        resultSpinner.setSelection(convertData.getTo());
    }

    static class SpinnerListener implements Spinner.OnItemSelectedListener {

        interface Callback {
            void convert();
        }

        private final Callback callback;

        private final TextView textView;

        SpinnerListener(TextView textView, Callback callback) {
            this.callback = callback;
            this.textView = textView;
        }

        @Override
        public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
            TextView childAt = (TextView) parent.getChildAt(0);
            if (childAt != null) childAt.setTextColor(Color.WHITE);

            String html = units.get(parent.getSelectedItemPosition()).getUnitSymbol();
            Spanned spanned = HtmlCompat.fromHtml(html, HtmlCompat.FROM_HTML_MODE_LEGACY);

            textView.setText(spanned);
            callback.convert();
        }

        @Override
        public void onNothingSelected(AdapterView<?> parent) {
            callback.convert();
        }
    }
}

File: app/src/main/java/org/nevack/unitconverter/converter/ConverterDisplayView.kt
package org.nevack.unitconverter.converter

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.app.Activity
import android.content.Context
import android.graphics.Rect
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroupOverlay
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.core.widget.doAfterTextChanged
import org.nevack.unitconverter.R
import org.nevack.unitconverter.converter.ConverterContract.ConvertData
import org.nevack.unitconverter.databinding.DisplayBinding
import org.nevack.unitconverter.model.ConversionUnit
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.sqrt

class ConverterDisplayView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0,
) : LinearLayout(context, attrs, defStyleAttr, defStyleRes) {
    private val binding: DisplayBinding
    private var conversionUnits: MutableList<ConversionUnit> = mutableListOf()

    private var sourceIndex = 0
    private var resultIndex = 0

    init {
        orientation = VERTICAL
        binding = DisplayBinding.inflate(LayoutInflater.from(context), this).apply {
            fab.setOnClickListener {
                val temp = sourceValue.text.toString()
                sourceValue.setText(resultValue.text.toString())
                resultValue.setText(temp)

                val newSourse = conversionUnits[resultIndex].name
                val newResult = conversionUnits[sourceIndex].name
                sourceSpinner.setText(newSourse, false)
                resultSpinner.setText(newResult, false)

                val tempIndex = resultIndex
                resultIndex = sourceIndex
                sourceIndex = tempIndex
            }
        }
    }

    fun setTextWatcher(watcher: (text: String) -> Unit) {
        binding.sourceValue.doAfterTextChanged {
            val text = it?.toString() ?: return@doAfterTextChanged
            watcher(text)
        }
    }

    fun showResult(result: String) {
        binding.resultValue.setText(result)
    }

    var convertData: ConvertData
        get() = with(binding) {
            ConvertData(
                sourceValue.text.toString(),
                resultValue.text.toString(),
                sourceIndex,
                resultIndex
            )
        }
        set(convertData) = with(binding) {
            sourceValue.setText(convertData.value)
            resultValue.setText(convertData.result)
            sourceIndex = convertData.from
            resultIndex = convertData.to
            sourceSpinner.setText(conversionUnits[sourceIndex].name, false)
            resultSpinner.setText(conversionUnits[resultIndex].name, false)
        }

    fun getCopyResult(withUnitSymbols: Boolean): String {
        return (binding.resultValue.text.toString()
                + if (withUnitSymbols) binding.sourceValueContainer.suffixTextView.text.toString() else "")
    }

    fun showError() {
        binding.resultValue.setText(R.string.message_error)
    }

    private fun setSpinnerAdapter(adapter: ArrayAdapter<String>) {
        binding.sourceSpinner.setAdapter(adapter)
        binding.sourceSpinner.setText(adapter.getItem(0), false)
        binding.sourceValueContainer.suffixText = conversionUnits[0].unitSymbol
        sourceIndex = 0

        binding.resultSpinner.setAdapter(adapter)
        binding.resultSpinner.setText(adapter.getItem(0), false)
        binding.resultValueContainer.suffixText = conversionUnits[0].unitSymbol
        resultIndex = 0
    }

    fun setUnits(conversionUnits: List<ConversionUnit>) {
        this.conversionUnits.clear()
        this.conversionUnits.addAll(conversionUnits)
        val adapter = ArrayAdapter(context,
            android.R.layout.simple_list_item_1,
            conversionUnits.map { it.name })
        setSpinnerAdapter(adapter)
    }

    fun erase() {
        val revealView = View(context)
        val revealAnimator: Animator
        val displayRect = Rect()
        this.getGlobalVisibleRect(displayRect)

        // Make reveal cover the display and status bar.
        revealView.bottom = displayRect.bottom
        revealView.left = displayRect.left
        revealView.right = displayRect.right
        revealView.setBackgroundColor(ContextCompat.getColor(context, R.color.colorAccent))
        val groupOverlay = (context as Activity).window.decorView.overlay as ViewGroupOverlay
        groupOverlay.add(revealView)
        var (width, height) = IntArray(2).apply(::getLocationInWindow)
        width += this.width
        height += this.height
        val revealCenterX = width - revealView.left
        val revealCenterY = height - revealView.top
        val x1 = (revealView.left - revealCenterX).toDouble().pow(2)
        val x2 = (revealView.right - revealCenterX).toDouble().pow(2)
        val y = (revealView.top - revealCenterY).toDouble().pow(2)
        val revealRadius = max(sqrt(x1 + y), sqrt(x2 + y)).toFloat()
        revealAnimator = ViewAnimationUtils.createCircularReveal(
            revealView,
            revealCenterX, revealCenterY, 0.0f, revealRadius
        )
        revealAnimator.duration =
            resources.getInteger(android.R.integer.config_longAnimTime).toLong()
        val alphaAnimator: Animator = ObjectAnimator.ofFloat(revealView, ALPHA, 0.0f)
        alphaAnimator.duration =
            resources.getInteger(android.R.integer.config_mediumAnimTime).toLong()
        alphaAnimator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                clear()
            }
        })
        val animatorSet = AnimatorSet()

        //Play Reveal Animation if Lollipop or higher, and only alpha animation for others
        animatorSet.play(revealAnimator).before(alphaAnimator)
        animatorSet.interpolator = AccelerateDecelerateInterpolator()
        animatorSet.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animator: Animator) {
                groupOverlay.remove(revealView)
            }
        })
        animatorSet.start()
    }

    fun clear() {
        binding.resultValue.editableText.clear()
        binding.sourceValue.editableText.clear()
    }

    fun appendText(text: String?) {
        binding.sourceValue.append(text)
    }

    fun removeLastDigit() {
        val textString = binding.sourceValue.text.toString()
        if (textString.isNotEmpty()) {
            binding.sourceValue.setText(textString.substring(0, textString.length - 1))
            binding.sourceValue.selectAll()
        }
    }

    fun setSpinnersCallback(callback: () -> Unit) {
        binding.sourceSpinner.setOnItemClickListener { _, _, position, _ ->
            sourceIndex = position
            binding.sourceValueContainer.suffixText = conversionUnits[position].unitSymbol
            callback()
        }
        binding.resultSpinner.setOnItemClickListener { _, _, position, _ ->
            resultIndex = position
            binding.resultValueContainer.suffixText = conversionUnits[position].unitSymbol
            callback()
        }
    }
}